pycode/memilio-generation/memilio/generation/ast.py

```diff
@@ -0,0 +1,154 @@
+        # file_args = file_args[1:-4] if len(file_args) > 5 else file_args
+
+        clang_cmd = [
+            "clang-14", self.config.source_file,
+            "-std=c++17", '-emit-ast', '-o', '-']
+        clang_cmd.extend(file_args)
+
+        try:
+            clang_cmd_result = subprocess.run(
+                clang_cmd, stdout=subprocess.PIPE)
+            clang_cmd_result.check_returncode()
+        except subprocess.CalledProcessError as e:
+            # Capture standard error and output
+            logging.error(
+                f"Clang failed with return code {e.returncode}. Error: {clang_cmd_result.stderr.decode()}")
+            raise RuntimeError(
+                f"Clang AST generation failed. See error log for details.")
+
+        # Since `clang.Index.read` expects a file path, write generated abstract syntax tree to a
+        # temporary named file. This file will be automatically deleted when closed.
+        with tempfile.NamedTemporaryFile() as ast_file:
+            ast_file.write(clang_cmd_result.stdout)
+            translation_unit = idx.read(ast_file.name)
+
+        self.assing_ast_with_ids(translation_unit.cursor)
+
+        logging.info("AST generation completed successfully.")
+
+        return translation_unit
+
+    def assing_ast_with_ids(self, cursor: Cursor) -> None:
+        """
+        Traverse the AST and assign a unique ID to each node during traversal.
+
+        @param cursor: The current node (Cursor) in the AST to traverse.
+        """
+        # assing_ids umschreiben -> mapping
+        self.cursor_id += 1
+        id = self.cursor_id
+        self.id_to_val[id] = cursor
+
+        if cursor.hash in self.val_to_id.keys():
+            self.val_to_id[cursor.hash].append(id)
+        else:
+            self.val_to_id[cursor.hash] = [id]
+
+        logging.info(
+            f"Node {cursor.spelling or cursor.kind} assigned ID {id}")
+
+        for child in cursor.get_children():
+            self.assing_ast_with_ids(child)
+
+    @property
+    def root_cursor(self):
+        return self.translation_unit.cursor
+
+    def get_node_id(self, cursor: Cursor) -> int:
+        """
+        Returns the id of the current node.
+        Extracs the key from the current cursor from the dictonary id_to_val
+
+        @param cursor: The current node of the AST as a cursor object from libclang.
+        """
+        for cursor_id in self.val_to_id[cursor.hash]:
+
+            if self.id_to_val[cursor_id] == cursor:
+
+                return cursor_id
+        raise IndexError(f"Cursor {cursor} is out of bounds.")
+
+    def get_node_by_index(self, index: int) -> Cursor:
+        """
+        Returns the node at the specified index position.
+
+        @param index: Node_id from the ast.
+        """
+
+        if index < 0 or index >= len(self.id_to_val):
+            raise IndexError(f"Index {index} is out of bounds.")
+        return self.id_to_val[index]
```

Comment:
Files should always end with an empty line
```suggestion
        return self.id_to_val[index]

```